### PR TITLE
[#PE-184] Enable Spid Access log

### PR DESCRIPTION
**Pull request by AleDore**
This PR is intended for enabling SPID access log. This should be the default behaviour, because of  AGID guidelines.

### CODE REVIEW

File: src/utils/blob.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  ).chain(fromEither);
+
+/**
+ * Get a blob content as a typed (io-ts) object.
+ *
+ * @param blobService     the Azure blob service
+ * @param containerName   the name of the Azure blob storage container
+ * @param blobName        blob file name
+ */
+export const getBlobAsObject = <A, S>(
```
Comment (BurnedMarshal) [inline src/utils/blob.ts:138]: Seems that this method is unused.

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: src/utils/blob.ts (hunk 1 above)
Comment (AleDore) [reply src/utils/blob.ts:138]: removed ;)

File: src/handlers/spid.ts [hunk 2]
```diff
@@ -21,3 +35,81 @@ export const metadataRefreshHandler = (
...
+      if (isNone(maybeFiscalCode)) {
+        logger.error(
+          `${logPrefix}|ERROR=Cannot get user's fiscal Code from SPID XML`
+        );
+        return;
+      }
+      const fiscalCode = maybeFiscalCode.value;
+
+      const errorOrSpidMsg = SpidLogMsg.decode({
+        createdAt: new Date(),
```
[the remark this answers is not in the document]
Comment (AleDore) [reply src/handlers/spid.ts:77]: yep it works, see (https://github.com/pagopa/io-backend/blob/5c1cdfd1e7af4487e25c2910f99c5fe29873f17e/src/utils/spid.ts#L112)[this]

File: src/handlers/spid.ts [hunk 3]
```diff
@@ -21,3 +35,81 @@ export const metadataRefreshHandler = (
...
+        );
+        return;
+      }
+      const spidMsg = errorOrSpidMsg.value;
+
+      logger.info(`Spid Msg => ${spidMsg.fiscalCode}`);
+
+      // we don't return the promise here
+      // the call follows fire & forget pattern
+      await storeSpidLogs(
```
[the remark this answers is not in the document]
Comment (AleDore) [reply src/handlers/spid.ts:99]: changed

File: src/handlers/spid.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/handlers/spid.ts:77]: Sorry, you have right. I see even the implementation of the validator [here](https://github.com/pagopa/io-ts-commons/blob/master/src/dates.ts#L41) and accept Date objects and ISO date strings

**Review by BurnedMarshal — APPROVED** [review]
LGTM

**Review by michaeldisaro — COMMENTED** [review]
LGTM, I'll help with tests and infra.

File: src/utils/__tests__/access_log.test.ts [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+  toEncryptedPayload: jest.fn((_, __) =>
+    right(({
+      cypherText: "encoded",
+      iv: "iv",
+      encryptedKey: "ek",
+    } as unknown) as encrypt.EncryptedPayload)
+  ),
+}));
+import * as encrypt from "@pagopa/ts-commons/lib/encrypt";
+const spiedToEncryptedPayload = jest.spyOn(encrypt, "toEncryptedPayload");
```
Comment (AleDore) [inline src/utils/__tests__/access_log.test.ts:39]: why not spy directly without jest.mock ? i.e.
```
const toEncryptedPayloadMock = jest.fn()
.mockImplementation((_,__) => right(({
      cypherText: "encoded",
      iv: "iv",
      encryptedKey: "ek",
    } as unknown) as encrypt.EncryptedPayload)
  );
jest.spyOn(encrypt, "toEncryptedPayload")
    .mockImplementation(toEncryptedPayloadMock);
```

File: src/utils/__tests__/access_log.test.ts [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+// Mock access_log module to spy on storeSpidLogs function
+jest.mock("../blob", () => ({
```
Comment (AleDore) [inline src/utils/__tests__/access_log.test.ts:42]: ditto

File: src/handlers/__tests__/spid.test.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+import { left, right } from "fp-ts/lib/TaskEither";
+import { some } from "fp-ts/lib/Option";
+import { toError } from "fp-ts/lib/Either";
+
+// Mock an express request and response
+const aMockedRequest = mockReq();
+const aMockedResponse = mockRes();
+
+// Mock access_log module to spy on storeSpidLogs function
+jest.mock("../../utils/access_log", () => ({
```
Comment (AleDore) [inline src/handlers/__tests__/spid.test.ts:27]: see comment on `access_log` test

File: src/utils/__tests__/access_log.test.ts (hunk 4 above)
Comment (michaeldisaro) [reply src/utils/__tests__/access_log.test.ts:39]: With this module it's not going to work. It's like we cannot mock just the spied function and it's not returning an either.

File: src/utils/__tests__/access_log.test.ts (hunk 5 above)
Comment (michaeldisaro) [reply src/utils/__tests__/access_log.test.ts]: Done in https://github.com/pagopa/hub-spid-login-ms/pull/28/commits/e9f29a1e8c4b53fa6f1b78857af9daef72bce9e7

File: src/handlers/__tests__/spid.test.ts (hunk 6 above)
Comment (michaeldisaro) [reply src/handlers/__tests__/spid.test.ts]: Done in https://github.com/pagopa/hub-spid-login-ms/pull/28/commits/e9f29a1e8c4b53fa6f1b78857af9daef72bce9e7

**Comment by AleDore** [comment]
It seems we need a rebase :)

**Review by michaeldisaro — APPROVED** [review]
LGTM

**Pull request merged**